File: javascript/queen-attack/queen-attack.js
//
// This is only a SKELETON file for the 'Queen Attack' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class QueenAttack {
  constructor({
    black: [blackRow, blackColumn] = [0, 3],
    white: [whiteRow, whiteColumn] = [7, 3],
  } = {}) {
    // Validate that queens are on the board (0-7 for both row and column)
    if (
      blackRow < 0 || blackRow > 7 || blackColumn < 0 || blackColumn > 7 ||
      whiteRow < 0 || whiteRow > 7 || whiteColumn < 0 || whiteColumn > 7
    ) {
      throw new Error('Queen must be placed on the board');
    }

    // Validate that queens don't occupy the same space
    if (blackRow === whiteRow && blackColumn === whiteColumn) {
      throw new Error('Queens cannot share the same space');
    }

    this.black = [blackRow, blackColumn];
    this.white = [whiteRow, whiteColumn];
  }

  toString() {
    // Create an 8x8 board with all empty spaces
    const board = Array(8).fill().map(() => Array(8).fill('_'));

    // Place the queens on the board
    const [whiteRow, whiteColumn] = this.white;
    const [blackRow, blackColumn] = this.black;
    
    board[whiteRow][whiteColumn] = 'W';
    board[blackRow][blackColumn] = 'B';

    // Convert the board to a string
    return board.map(row => row.join(' ')).join('\n');
  }

  get canAttack() {
    const [whiteRow, whiteColumn] = this.white;
    const [blackRow, blackColumn] = this.black;

    // Queens can attack if they are on the same row
    if (whiteRow === blackRow) {
      return true;
    }

    // Queens can attack if they are on the same column
    if (whiteColumn === blackColumn) {
      return true;
    }

    // Queens can attack if they are on the same diagonal
    // This happens when the absolute difference between rows equals the absolute difference between columns
    const rowDiff = Math.abs(whiteRow - blackRow);
    const colDiff = Math.abs(whiteColumn - blackColumn);
    
    return rowDiff === colDiff;
  }
}

File: cpp/queen-attack/queen_attack.cpp
#include "queen_attack.h"
#include <cmath>

namespace queen_attack {

chess_board::chess_board(std::pair<int, int> white_position, std::pair<int, int> black_position)
    : white_pos_(white_position), black_pos_(black_position) {
    
    // Validate both positions
    validate_position(white_pos_);
    validate_position(black_pos_);
    
    // Check that queens are in different positions
    if (white_pos_ == black_pos_) {
        throw std::domain_error("Queens cannot be placed in the same position");
    }
}

std::pair<int, int> chess_board::white() const {
    return white_pos_;
}

std::pair<int, int> chess_board::black() const {
    return black_pos_;
}

bool chess_board::can_attack() const {
    // Queens can attack if they are in the same row
    if (white_pos_.first == black_pos_.first) {
        return true;
    }
    
    // Queens can attack if they are in the same column
    if (white_pos_.second == black_pos_.second) {
        return true;
    }
    
    // Queens can attack if they are on the same diagonal
    // This happens when the absolute difference in rows equals the absolute difference in columns
    int row_diff = std::abs(white_pos_.first - black_pos_.first);
    int col_diff = std::abs(white_pos_.second - black_pos_.second);
    
    return row_diff == col_diff;
}

void chess_board::validate_position(const std::pair<int, int>& position) const {
    // Check that row and column are non-negative
    if (position.first < 0) {
        throw std::domain_error("Row position cannot be negative");
    }
    if (position.second < 0) {
        throw std::domain_error("Column position cannot be negative");
    }
    
    // Check that row and column are on the board (less than 8)
    if (position.first >= 8) {
        throw std::domain_error("Row position must be on the board (0-7)");
    }
    if (position.second >= 8) {
        throw std::domain_error("Column position must be on the board (0-7)");
    }
}

}  // namespace queen_attack
